File: ai/python/jaba-ai-python/neural_network.py
import json
import numpy as np

class NeuralNetwork:
    def __init__(self, inputnodes, hiddennodes, outputnodes, learningrate, wih=None, who=None):
        self.inputnodes = inputnodes
        self.hiddennodes = hiddennodes
        self.outputnodes = outputnodes
        self.learningrate = learningrate

        self.wih = wih if wih is not None else np.random.rand(hiddennodes, inputnodes) - 0.5
        self.who = who if who is not None else np.random.rand(outputnodes, hiddennodes) - 0.5

        self.act = np.vectorize(lambda x: 1 / (1 + np.exp(-x)))

    @staticmethod
    def normalize_data(data):
        return (data / 255) * 0.99 + 0.01

    def forward(self, input):
        input = input.reshape(-1, 1)
        if input.shape != (self.inputnodes, 1):
            raise ValueError("Input shape does not match expected shape.")
        h_in = self.wih @ input
        h_out = self.act(h_in)

        o_in = self.who @ h_out
        actual = self.act(o_in)

        self.cache = {'input': input, 'h_out': h_out, 'actual': actual}

        return actual

    def backward(self, target):
        target = target.reshape(-1, 1)

        if 'input' not in self.cache:
            self.cache['input'] = np.zeros((self.inputnodes, 1))

        h_in = self.wih @ self.cache['input']
        self.cache['h_out'] = self.act(h_in)

        o_in = self.who @ self.cache['h_out']
        self.cache['actual'] = self.act(o_in)

        dEdA = target - self.cache['actual']
        o_dAdZ = self.cache['actual'] * (1 - self.cache['actual'])

        dwho = (dEdA * o_dAdZ) @ self.cache['h_out'].T

        h_err = self.who.T @ (dEdA * o_dAdZ)
        h_dAdZ = self.cache['h_out'] * (1 - self.cache['h_out'])

        dwih = (h_err * h_dAdZ) @ self.cache['input'].T

        self.cache.update({'dwih': dwih, 'dwho': dwho, 'loss': np.sum(np.square(dEdA))})

    def update(self):
        if 'dwih' in self.cache and 'dwho' in self.cache:
            self.wih += self.learningrate * self.cache['dwih']
            self.who += self.learningrate * self.cache['dwho']
        else:
            print('Cannot update weights, some gradients are missing in the cache.')

    def load_weights(self, filename):
        with open(filename, 'r') as f:
            weights = json.load(f)

        self.wih = np.array(weights['wih'])
        self.who = np.array(weights['who'])

    def train(self, input, target):
        self.forward(input)
        self.backward(target)
        self.update()

    def predict(self, input):
        return self.forward(input)
    

File: ai/neural-network/feed-forward/NeuralNetwork.js
const { map } = require("lodash");
const math = require("mathjs");

/** these short cuts of mathJS functions make our life easier */
const mmap = math.map; // to be used to pass each element of a matrix to a function
const rand = math.random;
const transp = math.transpose;
const mat = math.matrix;
const e = math.evaluate;
const sub = math.subtract;
// const sqr = math.square;
const sum = math.sum;

function sqr(x) {
  if (Array.isArray(x)) {
    return x.map((val) => val * val);
  } else {
    return x * x;
  }
}

class NeuralNetwork {
  constructor(inputnodes, hiddennodes, outputnodes, learningrate, wih, who) {
    this.inputnodes = inputnodes;
    this.hiddennodes = hiddennodes;
    this.outputnodes = outputnodes;
    this.learningrate = learningrate;

    /* initialise the weights either randomly or, if passed in as arguments, with pretrained values */
    /* wih = weights of input-to-hidden layer */
    /* who = weights of hidden-to-output layer */
    this.wih = wih || sub(mat(rand([hiddennodes, inputnodes])), 0.5);
    this.who = who || sub(mat(rand([outputnodes, hiddennodes])), 0.5);

    /* the sigmoid activation function */
    this.act = (matrix) => mmap(matrix, (x) => 1 / (1 + Math.exp(-x)));
  }

  static normalizeData = (data) => {
    return data.map((e) => (e / 255) * 0.99 + 0.01);
  };

  cache = { loss: [] };

  forward = (input) => {
    const wih = this.wih;
    const who = this.who;
    const act = this.act;

    input = transp(mat([input]));

    /* hidden layer */
    const h_in = e("wih * input", { wih, input });
    const h_out = act(h_in);

    /* output layer */
    const o_in = e("who * h_out", { who, h_out });
    const actual = act(o_in);

    /* these values are needed later in "backward" */
    this.cache.input = input;
    this.cache.h_out = h_out;
    this.cache.actual = actual;

    return actual;
  };

  backward = (target) => {
    const who = this.who;
    const input = this.cache.input;
    const h_out = this.cache.h_out;
    const actual = this.cache.actual;

    target = transp(mat([target]));

    // calculate the gradient of the error function (E) w.r.t the activation function (A)
    const dEdA = sub(target, actual);

    // calculate the gradient of the activation function (A) w.r.t the weighted sums (Z) of the output layer
    const o_dAdZ = e("actual .* (1 - actual)", {
      actual,
    });

    // calculate the error gradient of the loss function w.r.t the weights of the hidden-to-output layer
    const dwho = e("(dEdA .* o_dAdZ) * h_out'", {
      dEdA,
      o_dAdZ,
      h_out,
    });

    // calculate the weighted error for the hidden layer
    const h_err = e("who' * (dEdA .* o_dAdZ)", {
      who,
      dEdA,
      o_dAdZ,
    });

    // calculate the gradient of the activation function (A) w.r.t the weighted sums (Z) of the hidden layer
    const h_dAdZ = e("h_out .* (1 - h_out)", {
      h_out,
    });

    // calculate the error gradient of the loss function w.r.t the weights of the input-to-hidden layer
    const dwih = e("(h_err .* h_dAdZ) * input'", {
      h_err,
      h_dAdZ,
      input,
    });

    this.cache.dwih = dwih;
    this.cache.dwho = dwho;
    this.cache.loss.push(sum(map(dEdA, sqr)));
  };

  update = () => {
    const wih = this.wih;
    const who = this.who;
    const dwih = this.cache.dwih;
    const dwho = this.cache.dwho;
    const r = this.learningrate;

    /* update the current weights of each layer with their corresponding error gradients */
    /* error gradients are negated by using the positve sign */
    this.wih = e("wih + (r .* dwih)", { wih, r, dwih });
    this.who = e("who + (r .* dwho)", { who, r, dwho });
  };

  predict = (input) => {
    return this.forward(input);
  };

  train = (input, target) => {
    this.forward(input);
    this.backward(target);
    this.update();
  };
}

module.exports = NeuralNetwork;
